pulsar-client-cpp/lib/BrokerConsumerStats.cc

```diff
@@ -0,0 +1,138 @@
+/**
+ * Copyright 2016 Yahoo Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <pulsar/BrokerConsumerStats.h>
+#include <lib/BrokerConsumerStatsImplBase.h>
+
+namespace pulsar {
+    BrokerConsumerStats::BrokerConsumerStats(boost::shared_ptr<BrokerConsumerStatsImplBase> impl) : impl_(impl) {}
+
+    BrokerConsumerStats::BrokerConsumerStats() {}
+
+    boost::shared_ptr<BrokerConsumerStatsImplBase> BrokerConsumerStats::getImpl() const {
+        return impl_;
+    }
+
+    bool BrokerConsumerStats::isValid() const {
+        if (impl_) {
+            return impl_->isValid();
+        }
+        return false;
+    }
+
+    std::ostream& operator<<(std::ostream &os, const BrokerConsumerStats& obj) {
+        os << "\nBrokerConsumerStats ["
+           << "validTill_ = " << obj.isValid()
+           << ", msgRateOut_ = " << obj.getMsgRateOut()
+           << ", msgThroughputOut_ = " << obj.getMsgThroughputOut()
+           << ", msgRateRedeliver_ = " << obj.getMsgRateRedeliver()
+           << ", consumerName_ = " << obj.getConsumerName()
+           << ", availablePermits_ = " << obj.getAvailablePermits()
+           << ", unackedMessages_ = " << obj.getUnackedMessages()
+           << ", blockedConsumerOnUnackedMsgs_ = " << obj.isBlockedConsumerOnUnackedMsgs()
+           << ", address_ = " << obj.getAddress()
+           << ", connectedSince_ = " << obj.getConnectedSince()
+           << ", type_ = " << obj.getType()
+           << ", msgRateExpired_ = " << obj.getMsgRateExpired()
+           << ", msgBacklog_ = " << obj.getMsgBacklog()
+           << "]";
+        return os;
+    }
+
+    double BrokerConsumerStats::getMsgRateOut() const {
+        if (impl_) {
```

Comment:
it would be more idiomatic to just document that no getXXX is allowed on an invalid BrokerConsumerStats object and get rid of the check. that avoids hiding bugs.